### PR TITLE
Added tutorial on Lazy expressions to tutorial section of manual

### PR DESCRIPTION
**Pull request by ulugbekna**
Added a tutorial on Lazy expressions to the tutorial section of the manual. [MPR#7547](https://caml.inria.fr/mantis/view.php?id=7547)

Content based on:

1. [standard-library module Lazy](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Lazy.html)
2. [Lazy patterns docs](https://github.com/yallop/ocaml-patterns/wiki/Lazy-patterns) by @yallop 
3. Own experience and tests

I was also hesitant whether to make the link to module Lazy docs open in a new tab or the same. The former seems better UX. 

Initially, I also wanted to add an example with a recursive fibonacci function call with a large value that would exemplify an expensive computation but decided not to for brevity.

### CODE REVIEW

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: You could add a small side-effect to the expression, to make it clearer when the expression is evaluated. For instance,
```OCaml
let lazy_two = lazy ( print_endline "lazy_two evaluation"; 1 + 1 )
```

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: With a small side-effect added to lazy_two, it may be simpler to check that
```OCaml
let lazy x = lazy_two
```
do not print "Lazy_two evaluation" anymore.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: I fear that this paragraph plays a bit too fast: it goes from a grammatical description to a technical detail relatively fast:

In general, I am not sure if the grammar description "| lazy <pattern>" fits this chapter (moreover the leading `|` should be removed).

Alternatively, I would start the paragraph by stating that "Lazy patterns provide another way to force a lazy expression"
Then, a simpler illustration for lazy pattern might be:

```OCaml
    let lazy_l = lazy ([1;2] @ [3;4])
    let lazy l = lazy_l;;
```
Chaining that example with nested lazy pattern example would be a good stepping stone.
And finally, at this point,  it would make sense to tell our readers that a lazy pattern always force the evaluation of the deferred computation.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: Is "Pay attention" not a bit strong?
The redundancy `plain, unboxed` seems unnecessary and potentially confusing since there is another notion of unboxed.
Maybe just
> Notice that our function call returns a plain value

?

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: It may be worth it to be a bit more plain here (otherwise, constant time is both technical and not that clear) . Maybe something like

> In other words, every subsequent call of "Lazy.force" returns the result of the first computation
without recomputing the lazy expression.

?

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: (Note that this link only works in the  html version of the manual, but we can fix this point later)

**Review by Octachron — COMMENTED** [review]
Thanks for your contribution! This looks nice, except for the lazy pattern part that might need to be a bit more gradual. (I have also some remarks on potential improvement earlier but you can ignore those ones).

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: Thanks, that's a good idea.

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: This snippet comes before I introduce lazy patterns, i.e., `let lazy x = ...`, so users might not know this syntax (?).  What do you think if I simply use `Lazy.force` second time to show that `print_endline` and thus the whole expression is _not_ run again?

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: Agreed. I had similar thoughts about "plain, unboxed".

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: What other versions of the manual do you have in mind? 
I feel that it would be best for newcomers if we resolved this issue asap. Not working links in docs cause frustration.

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: Your version seems easier to understand 👍

File: manual/manual/tutorials/coreexamples.etex
Comment (gasche) [reply manual/manual/tutorials/coreexamples.etex]: Some people read the PDF version of the manual.

As @Octachron suggested in the comment above, feel free to left this particular issue unchanged in your PR. @Octachron may be working on a convenient macro to add links to the standard library in a portable way, but for now we can leave the HTML link.

On the other hand, it would be nice if you could add more commits or rebase to address the other issues in the rest of the review.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [reply manual/manual/tutorials/coreexamples.etex]: Reusing `Lazy.force` is indeed probably better at this point.

**Comment by ulugbekna** [comment]
I have taken into account comments by @Octachron and fixed some parts of the tutorial. Thanks :)
Also did some changes in other places, where I felt they were due.

**Comment by gasche** [comment]
(I trust @Octachron's judgment and review, and I think he will merge the PR once he declares himself satisfied.)

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: I think that the emphasis on the internal status on  `lazy_t` is a bit strong.
We could start with the second point 

> the value of "lazy_two" is "<lazy>"  …

Then, the first point could be a short note, maybe even a footnote. It could be also shortened along the line of

> Note that "lazy_two" has type "int lazy_t". However, the type "'a lazy_t" is an internal type name, when possible the type `'a Lazy.t` should be preferred.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex:650]: I think it would make the example stronger, if the `lazy _` pattern really forced a computation. Currently, in this example, the computation is already forced by `lazy 0`. A possibility might be
```OCaml
let maybe_eval guard lazy_expr = match guard, lazy_expr with
| lazy false, _ -> "matches if (Lazy.force guard = false)"
| lazy true, lazy _ -> "despite being a lazy wildcard pattern, always forces the lazy expr";;
```
Then the explanation below could mention that the lazy expression is forced only if the `guard` value yields `true` once computed.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: If possible I prefer to avoid starting sentence with a variable names like `force`. For instance, the above could be rewritten as 

> The function "force" comes from

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: I think we could be more straightforward and tell the reader that the line `lazy_two evaluated` is not printed anymore: like this the reader doesn't have to remember what they could have seen if  `print_endline` had been called.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: Maybe just ` "2"` rather than `i.e. "2"`?

**Review by Octachron — COMMENTED** [review]
I sill have few more remarks, but this is definitively converging towards a nice addition to the manual.

File: manual/manual/tutorials/coreexamples.etex
Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: I tried to change it to simply `"2"`, but it reads "the result of that expression 2", doesn't it?

Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: as if our expression is 2 not 1+1

Comment (Octachron) [reply manual/manual/tutorials/coreexamples.etex]: Well, the comma would still be here:

> the result of that expression, "2".

or an em dash (`---` in latex) could work too:

> the result of that expression—"2".

But `i.e.` is fine too
> the result of that expression, i.e. "2".

This is merely a matter of taste, you can ignore my remark if you wish.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: The two sentences could be connected:
```
The lazy expression "lazy_expr" is forced only if the "lazy_guard" value yields 
"true" once computed.  Indeed, a simple wildcard pattern (not lazy) never forces
the lazy expression's ...
```

**Comment by Octachron** [comment]
This looks nice, I will do a last round of review tomorrow (for spelling and typos mostly).

File: manual/manual/tutorials/coreexamples.etex
Comment (gasche) [inline manual/manual/tutorials/coreexamples.etex]: `is displayed as "<lazy>"` (this is not the actual value)

File: manual/manual/tutorials/coreexamples.etex
Comment (gasche) [inline manual/manual/tutorials/coreexamples.etex]: > The value `"lazy_two"` is not displayed as "<lazy>" anymore but "lazy 2".

**Review by gasche — APPROVED** [review]
I made a minor nitpick, but with it the PR is good to go -- I'll merge when it is fixed.

File: manual/manual/tutorials/coreexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/coreexamples.etex]: Final nitpick: maybe more `final value` than `actual value`?

Comment (ulugbekna) [reply manual/manual/tutorials/coreexamples.etex]: Yes, that makes sense because "actual value" might also pertain to value displayed as <lazy>.

**Review by Octachron — APPROVED** [review]
LGTM

**Comment by Octachron** [comment]
@ulugbekna , before me (or @gasche)  merge, would you mind adding a change entry in `Changes` (in a new ` ### Manual and documentation:` subsection in the `Working version` section)? The entry format is described in see [Contributing.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#changelog).

**Comment by ulugbekna** [comment]
@Octachron, thanks for a heads-up! :-)

**Comment by Octachron** [comment]
Could you add the GPR number to the change entry (i.e. MPR#7447, GPR#2273)?

@gasche , any objection to cherry-picking to 4.08?

**Comment by gasche** [comment]
@Octachron none, especially if you do the work of merging and cherry-picking yourself. Thanks!

**Comment by Octachron** [comment]
Merged and cherry-picked on 4.08 as  d11878cf70d2a9f.
@ulugbekna , thanks a lot for your contribution!

**Comment by ulugbekna** [comment]
@Octachron and @gasche, thank you a lot for guiding through my first contribution. I appreciate your help and reviews :-)

**Pull request merged**
